Refactor Event Update Ordering

fmatthew5876 opened the pull request:
The PR refactors a number of interpreter and event related items

* Adds support for `triggered_by_decision_key` chunk
* Adds support for `pause` chunk
* Adds support for `waiting_execution` chunk
* Adds support for `processed` chunk
* Fixes the execution order of event update routines to match RPG_RT.
* Correctly handles "PreUpdate" which is the partial update routine done at startup which only updates events but not player or vehicles. This also happens when you teleport or use escape/teleport from the menu. *I haven't addressed the teleport PreUpdate() cases yet and may not do so in this PR*.

Fix: #1623 
Fix: #1247
Fix: #1595 
Fix: #663
Fix: #1664

Breaks: #937

Ghabry commented:
Jenkins: Test this please

fmatthew5876 commented:
This PR is now becoming very experimental as I refactor how and when the interpreter starts. Please don't recommend people to run it in continuous builds until we have a good suite of passing test cases.

fmatthew5876 commented:
This PR is currently being redone. The problem was I had many commits attempting to to figure out `trigger=touch` and `trigger=collision` logic. Trying one thing, undoing it for another, etc.. Many useless commits.

This is reorganizing the other stuff (pause, processed, set waiting flag and run interpreter later, make way, etc..) first. And then I will redo the triggering logic last.

Ghabry commented:
I would prefer to have everything starting from "Refactor MakeWay" moved to another PR because one extremely complex interpreter commit ("Refactor how foreground interpreter is triggered") per PR is more than I feel comfortable to handle.

fmatthew5876 commented:
Sure, but in that incomplete state I'm not sure if we cause more regressions.

I'll reorganize the PRs soon.

fmatthew5876 commented:
This is ready for review.

All test cases pass in #1664 except:
- [ ] 1.3 - The pause and waiting flags disapear on the 4th save instead of the 3rd. This is likely a timing issue for when exactly we change event pages. If I get to this soon maybe I'll put it in this PR, or save it for later.

fmatthew5876 commented:
Now that we have #1689 I've reintroduced the "Remove IsBlockedByMoveRoute()" commit from #1601 as that's really not needed anymore.

Ghabry commented:
This PR breaks faster decrement of the Wait counter when an interpreter is preempted when walking against events. I also checked against #1687 with same result.

Have two parallel events on same-as-hero-layer:

``
@> Wait: 5.0 seconds
@> Text: A

@> Wait: 5.0 seconds
@> Text: B
``

Walk against one of the events the whole time, this event will show the message after 2.5 seconds.
(This can be fixed in a later PR, as this is a quite obscure "feature")

Ghabry commented:
I can confirm that this really breaks Pom Gets Wifi, so the Pom-Bug is actually two problems in one: Movability while a move route is active + this wait bug.

So to track this will simply reopen #937 when this gets merged.

Ghabry commented:
"Check for map refresh before exiting **intpreter** Update()"

Ghabry commented on src/game_player.cpp:
```diff
@@ -202,6 +200,16 @@ void Game_Player::UpdateSelfMovement() {
+			if (force_through) {
+				SetThrough(true);
+			}
+			auto sg = makeScopeGuard([&](){
```
is the usage of a scope guard really needed here?
There is only one exit path and we don't throw/catch exceptions.
Though what actually bugs me is that this comes from liblcf. We already have other borderline-stuff in liblcf (INI) we call from Player but imo here the header should be copied into Player and made a private/not-installed header in liblcf.

fmatthew5876 replied:
I always get paranoid whenever there is a thing like this where you temporarily set some state, call a function, and then restore it. This gets optimized away and we ever change the code later to throw it's still safe.

Overly paranoid for sure but that's usually how I do these things as a habit. Can take it away if its really objectionable.

Ghabry replied:
I understand the intention but I want to discourage that liblcf becomes a general collection of convenience headers that are used by Player because we don't use C++17.
So I would be already happy if you put a copy of scopeguard into the Player src folder, too.

Ghabry commented on src/game_event.cpp:
```diff
@@ -553,58 +549,24 @@ void Game_Event::Update() {
-	// Placed after the interpreter update because multiple updates per frame are allowed.
-	// This results in waits to finish quicker when an event collides with this event and
-	// emulates a RPG Maker bug)
-	// Only update the event once per frame
```
just marking this here because this comment must be readded when this is fixed again

```
// Placed after the interpreter update because multiple updates per frame are allowed.
// This results in waits to finish quicker when an event collides with this event and 
// emulates a RPG Maker bug) 
```

Ghabry commented on src/player.cpp:
```diff
@@ -306,7 +306,11 @@ void Player::Update(bool update_scene) {
 	for (int i = 0; i < speed_modifier; ++i) {
+		bool was_transition_pending = Graphics::IsTransitionPending();
 		Graphics::Update();
+		if (was_transition_pending && !Graphics::IsTransitionPending() && Scene::instance->IsInitialized()) {
```
I don't find this intuitive. Could you add a comment to explain that this is used by Scene Map to handle scene calling before the first update?

fmatthew5876 replied:
I don't find it intuitive at all. In fact I hate it, but I don't know how else to solve this.

I will add the comment.

Ghabry commented:
Well this PR was much easier to review than the STop Count one but MakeWay will be hell again

fmatthew5876 commented:
I've added #1724 to discuss `ScopeGuard` and other utilities. We can follow up with another PR whatever we decide there.

For the wait bug, we can reopen #937 and put details in there until its fixed again after #1687.

I've addressed the other comments.

carstene1ns commented on src/game_commonevent.cpp:
```diff
@@ -109,3 +88,17 @@ RPG::SaveEventData Game_CommonEvent::GetSaveData() {
+	return ce->trigger == RPG::EventPage::Trigger_parallel &&
+		(!ce->switch_flag || Game_Switches.Get(ce->switch_id))
+		&& !ce->event_commands.empty();
+}
```
Maybe these could be combined?

fmatthew5876 replied:
Sure, I'll drop a commit for that into the next one #1687

carstene1ns commented on src/game_interpreter.cpp:
```diff
@@ -142,11 +142,18 @@ void Game_Interpreter::SetContinuation(Game_Interpreter::ContinuationFunction fu
 }
 
+bool Game_Interpreter::ReachedLoopLimit() const {
+	return loop_count >= 10000;
```
This should be a constant here already. I have seen this in another commit.

fmatthew5876 replied:
Thats fixed in #1713

carstene1ns commented on src/game_event.cpp:
```diff
@@ -315,45 +315,41 @@ bool Game_Event::GetActive() const {
+		return;
 	}
 }
 
```
These returns seem unnecessary now.

fmatthew5876 replied:
Will fix in #1687 along with the previous comment

fmatthew5876 replied:
This call gets blown away in #1697 so I'm leaving it alone for now.
